src/ring_buffer.zig

```diff
@@ -20,54 +23,74 @@ pub fn RingBuffer(comptime T: type, comptime size: usize) type {
         }
 
         /// Return but do not remove the next item, if any.
-        pub fn peek(self: *Self) ?T {
+        pub fn peek(self: *Self) ?*T {
             if (self.empty()) return null;
-            return self.buffer[self.index];
+            return &self.buffer[self.index];
         }
 
         /// Remove and return the next item, if any.
         pub fn pop(self: *Self) ?T {
             if (self.empty()) return null;
-            const ret = self.buffer[self.index];
-            self.index = (self.index + 1) % self.buffer.len;
-            self.count -= 1;
-            return ret;
+            defer {
+                self.index = (self.index + 1) % self.buffer.len;
+                self.count -= 1;
+            }
+            return self.buffer[self.index];
         }
 
-        pub fn full(self: Self) bool {
+        /// Returns whether the ring buffer is completely full.
+        pub fn full(self: *const Self) bool {
             return self.count == self.buffer.len;
         }
 
-        pub fn empty(self: Self) bool {
+        /// Returns whether the ring buffer is completely empty.
+        pub fn empty(self: *const Self) bool {
             return self.count == 0;
         }
+
+        pub const Iterator = struct {
+            ring: *Self,
+            count: usize = 0,
+
+            pub fn next(it: *Iterator) ?*T {
+                assert(it.count <= it.ring.count);
+                if (it.count == it.ring.count) return null;
+                defer it.count += 1;
+                return &it.ring.buffer[(it.ring.index + it.count) % it.ring.buffer.len];
+            }
+        };
+
+        /// Returns an iterator to iterate through all `count` items in the ring buffer.
+        pub fn iterator(self: *Self) Iterator {
```

Comment:
This could probably use a comment stating that `push()`/`pop()` while iterating are not safe.